packages/flutter_test/lib/src/binding.dart

```diff
@@ -936,41 +936,83 @@ class AutomatedTestWidgetsFlutterBinding extends TestWidgetsFlutterBinding {
       );
     }());
 
+    final List<Timer> timerCache = <Timer>[];
+    final List<Timer> repeatingTimerCache = <Timer>[];
+    int microtasks = 0;
+    DateTime latestMicrotaskCreationTime;
     final Zone realAsyncZone = Zone.current.fork(
       specification: ZoneSpecification(
         scheduleMicrotask: (Zone self, ZoneDelegate parent, Zone zone, void f()) {
-          Zone.root.scheduleMicrotask(f);
+          print('1 with ${self.hashCode}');
+          microtasks++;
+          latestMicrotaskCreationTime = DateTime.now();
+          Zone.root.scheduleMicrotask(() async {
+            f();
+            microtasks--;
+            print('Done');
+          });
         },
         createTimer: (Zone self, ZoneDelegate parent, Zone zone, Duration duration, void f()) {
-          return Zone.root.createTimer(duration, f);
+          print('2 with $duration');
+          final Timer timer = Zone.root.createTimer(duration, f);
+          timerCache.add(timer);
+          return timer;
         },
         createPeriodicTimer: (Zone self, ZoneDelegate parent, Zone zone, Duration period, void f(Timer timer)) {
-          return Zone.root.createPeriodicTimer(period, f);
+          print(3);
+          final Timer timer = Zone.root.createPeriodicTimer(period, f);
+          repeatingTimerCache.add(timer);
+          return timer;
         },
       ),
     );
 
     addTime(additionalTime);
 
-    return realAsyncZone.run<Future<T>>(() {
+    return realAsyncZone.run<Future<T>>(() async {
       _pendingAsyncTasks = Completer<void>();
-      return callback().catchError((dynamic exception, StackTrace stack) {
+      final Completer<bool> sideEffectCompleter = Completer<bool>();
+      const Duration checkInterval = Duration(milliseconds: 50);
+      Zone.root.createPeriodicTimer(checkInterval, (Timer timer) {
+        timerCache.retainWhere((Timer timer) => timer.isActive);
+        repeatingTimerCache.retainWhere((Timer timer) => timer.isActive);
+        print('Checking $microtasks $timerCache $latestMicrotaskCreationTime');
+        if ((latestMicrotaskCreationTime == null || DateTime.now().difference(latestMicrotaskCreationTime).inMilliseconds > 500)
+            && microtasks == 0
+            && timerCache.isEmpty
+            && repeatingTimerCache.isEmpty) {
+          print('Success!');
+          timer.cancel();
+          sideEffectCompleter.complete(true);
+        } else if ((checkInterval * (timer.tick - 10)) >= additionalTime) {
```

Comment:
The purpose of additional time is to increase the timeout value for the test section, I think we should not overload its usage. As of the current implementation it both increases the wait time for the timer/microtask and also extends the timeout.

It also raise another question, what if there are unresolved futures that user does not care about? currently it will always fail if there are unresolved futures after the additionalTime.

How about we expose a new api, something like `tester.resolveTimerAndMicrotask()`. this api create a periodic timer just like you do here, but the sideEffectCompleter will only resolve when there are no more timercache and microtasks.
In the test, the developer can decide if they want to wait for the future by calling `await tester.resolveTimerAndMicrotask()`. If they have a bad test that creating unresolved timer or microtask, it will just hang.

We don't do any waiting at the runAsync tear down, it will be developers choices to decide what they want to do.

Reply:
> The purpose of additional time is to increase the timeout value for the test section, I think we should not overload its usage. As of the current implementation it both increases the wait time for the timer/microtask and also extends the timeout.

Fair point. I can introduce a new parameter and also use this parameter for signaling that runAsync() should not wait for additional futures.

```dart
waitForIdle: const Duration(seconds: 2),
```
I would rather not introduce a new API. I would expect, in the normal case, for users to actually care about the Futures that their widget code introduces in runAsync. They would not be using runAsync otherwise.  We should optimize for that most common case. Once this solution makes it to SO, everyone will simply copy and use them together anyway. The APIs being separate will lose its meaning.

I agree with you that there will be advanced cases where these futures should not be awaited but then we are getting into highly custom test code category. Users can then create a future that they want to await and do that. I will put all this in actionable error messages.

```dart
// Do not wait for idle.
await runAsync(() => some code, waitForIdle: null);
await runAsync(() => await someSpecialFuture);
```

Another reason: The interplay between the two APIs would need to be carefully covered. For instance what if this happens:

```dart
await runAsync;
await resolveTimers; // What if runAsync runs into error. Now the user needs to check return value and make sure there was no error in the async block.
//
// some code later
//
await resolveTimers;  // Does this throw? Does it just return? Either option is error prone.
```

Let me know what you think.

Reply:
I think i agree with most of it, but there are some more concerns i have.

> waitForIdle: const Duration(seconds: 2),

I think we don't need the duration here. this can be a true false value, and the test will either resolve the timer/mt or just timeout the test section.

>I agree with you that there will be advanced cases where these futures should not be awaited but then we are getting into highly custom test code category. Users can then create a future that they want to await and do that. I will put all this in actionable error messages.

I think part of this is also we have some unexposed future in our built-in widget as well (for example the image decoding in Image widget or localization asset loading). unless we expose those future in the widget state, user will not be able to resolve them If their test the result after those async load. That is why I kinda want to have a separate api to resolve this future. However, I am ok with not provide an API too since we can use the runAsync. waitForIdle to resolve the future. It just the test code maybe a bit more complicated.

>Another reason: The interplay between the two APIs would need to be carefully covered. For instance what if this happens:

I will expect resolveTimers called without an async zone (not in the runAsync callback) will just throw an error?

Reply:
I see what you mean now. I thought you were proposing a separate API to await *after* runAsync was done.

Yes, providing a helper function to call inside runAsync callback is tempting. I still prefer the default behavior to be more robust because when things fail today, they fail without giving developers any direction on what might be the problem.

That's why I still prefer having the default be more robust and in the case of failure, signaling that the user needs to define a custom task future to await inside runAsync().

Reply:
I will think about this a bit more. Running this PR over google3 broke 24 tests. At least 2 of them were pre-existing bugs in the tests themselves that this uncovered! However, many are due to premature timeouts. Perhaps I will have to do what you are proposing after all @chunhtai and expose the periodic timer as a `Future` for apps to listen on. I will dive deep into some use cases and see what I missed. The good news is there are at least 9000 tests that exercise realAsync and only 24 failed.

Reply:
Latest results: 10 failures.

- 2 due to pre-existing bugs in tests.
- 4 due to the tested code creating a periodic timer that never resolves.
   - I am inclined to warn people about this rather than awaiting on them.
- 4 due to running *a lot* of tests in a single file. This is the problematic one. Since we now wait up to 500ms to make sure there are no new microtasks created, each test (that uses `runAsync`) is lengthened by half a second which means if you are running 120+ tests, your test now runs at least a minute longer.

Due to #2 and #3, I am now leaning towards your proposal to expose this as a separate method. It would be nice to make the default method more robust but it should not come at an expense of causing half a second delay per test.

I will put in some robust documentation as well so people can make an informed decision on when to use this API.